Add support for the out-of-tree Marvell Prestera driver

jpetrina opened the pull request:
Pushing support for the out-of-tree Marvell Prestera driver [1] for review.

This is based on https://github.com/dentproject/meta-switchbsp/pull/1, https://github.com/dentproject/meta-switchbsp/pull/2, and https://github.com/dentproject/meta-switchbsp/pull/3.

This change generates the following packages:

```
109K  build/tmp/deploy/ipk/delta_tn48m/kernel-module-prestera-5.15.11-yocto-standard_3.2.2-r0_delta_tn48m.ipk
2.5M  build/tmp/deploy/ipk/delta_tn48m/kernel-module-prestera-dbg_3.2.2-r0_delta_tn48m.ipk
1012  build/tmp/deploy/ipk/delta_tn48m/kernel-module-prestera-dev_3.2.2-r0_delta_tn48m.ipk
127K  build/tmp/deploy/ipk/delta_tn48m/kernel-module-prestera-src_3.2.2-r0_delta_tn48m.ipk
558   build/tmp/deploy/ipk/delta_tn48m/kernel-module-prestera_3.2.2-r0_delta_tn48m.ipk
```

[1] https://github.com/Marvell-switching/switchdev-prestera

cc @damirsamardzic

Commit:
initial layer structure

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
README.md: add testing instructions

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf: layer: remove unused dynamic layers definitions

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf: machine: arm64-all: remove unused NXP RDBs

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
README: fix link identation level

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf: layer: add missing meta-oe layer as dependency

The README mentions this layer as a dependency so enforce this in the
build system.

Reported-by: Jonas Gorski <jonas.gorski@bisdn.de>
Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
recipes-kernel: linux: extend default linux-yocto with switch-kmeta

This change introduces a new kernel metadata structure with generic
arm64 and x86_64 definitions for switches. This can be extended later on
with platform-specific overrides.

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf/machine: move kernel defs to individual machine

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
*: add import-platforms script

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf/machine: add initial selected platforms from SONiC and ONL

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
recipes-kernel: linux-dentproject: temporarily skip qa before pulling upstream fixes

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf: delta-tn48m: add SERIAL_CONSOLES definition to populate inittab

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
recipe-images: add core-image-minimal-onie package

This change adds a simple package which builds on top of
`core-image-minimal` to provide ONIE installer image which installs the
rootfs onto device's internal hard drive.

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
recipes-kernel: add kernel-module-prestera recipe for building out-of-tree

This change adds a new recipe which builds an out-of-tree version of
the Marvell swichdev driver for Prestera.

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
recipes-images: core-image-minimal-onie: fix multiline string definition

This resolves a warning in the build system regarding inability to parse
the DESCRIPTION variable.

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

Commit:
conf: delta-tn48m: build kernel-module-prestera out-of-tree driver by default

After the kernel-module-prestera recipe is introduced, we disable the
same in-kernel module in the default configuration and enable the new
out-of-tree module.

Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

taraschornyiplv commented:
got this error

```
Loaded 3282 entries from dependency cache.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "2.9.1"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-poky-linux"
MACHINE              = "delta-tn48m"
DISTRO               = "poky"
DISTRO_VERSION       = "5.1"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "master:b9b59b5a8bdfa2666477018e7175c5036376807c"
meta-oe              = "master:01dd622533957d20fa860e34ff098161a103bfbf"
meta-switchbsp       = "develop-outoftree-prestera:c0ed18139c9d3a8844602270b008ea203829346b"

Sstate summary: Wanted 4 Local 0 Mirrors 0 Missed 4 Current 1796 (0% match, 99% complete)#################################################################               | ETA:  0:00:00
Initialising tasks: 100% |###############################################################################################################################################| Time: 0:00:02
NOTE: Executing Tasks
ERROR: core-image-tiny-initramfs-1.0-r0 do_rootfs: Could not invoke dnf. Command '/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/recipe-sysroot-native/usr/bin/dnf -v --rpmverbosity=info -y -c /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/etc/dnf/dnf.conf --setopt=reposdir=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/etc/yum.repos.d --installroot=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs --setopt=logdir=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp --repofrompath=oe-repo,/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/oe-rootfs-repo --nogpgcheck install base-passwd busybox busybox-mdev dropbear initramfs-live-boot-tiny packagegroup-core-boot run-postinsts' returned 1:
DNF version: 4.21.1
cachedir: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/var/cache/dnf
Added oe-repo repo from /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/oe-rootfs-repo
User-Agent: falling back to 'libdnf': could not detect OS or basearch
repo: using cache for: oe-repo
oe-repo: using metadata from Fri 15 Nov 2024 09:47:21 PM UTC.
Last metadata expiration check: 0:00:01 ago on Wed 04 Dec 2024 03:21:01 PM UTC.
No match for argument: busybox-mdev
Error: Unable to find a match: busybox-mdev

ERROR: Logfile of failure stored in: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp/log.do_rootfs.5850
ERROR: Task (/home/taras/git/poky/meta/recipes-core/images/core-image-tiny-initramfs.bb:do_rootfs) failed with exit code '1'
NOTE: Tasks Summary: Attempted 3720 tasks of which 3719 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/taras/git/poky/meta/recipes-core/images/core-image-tiny-initramfs.bb:do_rootfs
    log: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp/log.do_rootfs.5850

```

jpetrina commented:
Hi @taraschornyiplv, thanks for the feedback.

I think this issue is frequent when there isn't any `VIRTUAL-RUNTIME_dev_manager` defined by the user. Basically the target expects some kind of device manager implementation to be pulled in the image.

> got this error
> 
> ```
> ...
> DISTRO               = "poky"
> ...
> ERROR: core-image-tiny-initramfs-1.0-r0 do_rootfs: Could not invoke dnf. Command '/home/
> ...
> ```

So this is the `core-image-tiny{,-initramfs}` target with `DISTRO=poky`, could you try building this target with `DISTRO=poky-tiny`?

Alternatively, you could try another target e.g. the README suggests `core-image-minimal-initramfs`.

Otherwise, you could also use https://github.com/dentproject/meta-dentos/pull/1 which provides `DISTRO=dentos` which pulls in systemd implementations of init and dev manager.

taraschornyiplv commented:
when  used poky-tiny  

```taras@taras-chornyi-pc:~/git/poky/build$ bitbake core-image-minimal-initramfs
Loading cache: 100% |                                                                                                                                                   | ETA:  --:--:--
Loaded 0 entries from dependency cache.
Parsing recipes: 100% |##################################################################################################################################################| Time: 0:00:07
Parsing of 1904 .bb files complete (0 cached, 1904 parsed). 3282 targets, 554 skipped, 0 masked, 0 errors.
Removing 93 recipes from the armv8a sysroot: 100% |######################################################################################################################| Time: 0:01:01
Removing 102 recipes from the delta_tn48m sysroot: 100% |################################################################################################################| Time: 0:00:04
NOTE: Resolving any missing task queue dependencies
ERROR: Nothing PROVIDES 'virtual/kernel'
linux-yocto-tiny PROVIDES virtual/kernel but was skipped: incompatible with machine delta-tn48m (not in COMPATIBLE_MACHINE)
linux-yocto-upstream PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-yocto-upstream to enable it
linux-yocto PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-yocto to enable it
linux-yocto-upstream PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-yocto-upstream to enable it
linux-yocto PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-yocto to enable it
linux-yocto-rt PROVIDES virtual/kernel but was skipped: incompatible with machine delta-tn48m (not in COMPATIBLE_MACHINE)
linux-yocto-tiny PROVIDES virtual/kernel but was skipped: incompatible with machine delta-tn48m (not in COMPATIBLE_MACHINE)
linux-dummy PROVIDES virtual/kernel but was skipped: PREFERRED_PROVIDER_virtual/kernel set to linux-yocto-tiny, not linux-dummy
linux-yocto-dev PROVIDES virtual/kernel but was skipped: incompatible with machine delta-tn48m (not in COMPATIBLE_MACHINE)
linux-yocto-rt PROVIDES virtual/kernel but was skipped: incompatible with machine delta-tn48m (not in COMPATIBLE_MACHINE)
linux-dentproject PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-dentproject to enable it
ERROR: Required build target 'core-image-minimal-initramfs' has no buildable providers.
Missing or unbuildable dependency chain was: ['core-image-minimal-initramfs', 'virtual/kernel']

```

jpetrina commented:
Thanks,

> when used poky-tiny
> 
> ```
> ...
> linux-dentproject PROVIDES virtual/kernel but was skipped: Set PREFERRED_PROVIDER_virtual/kernel to linux-dentproject to enable it
> ERROR: Required build target 'core-image-minimal-initramfs' has no buildable providers.
> Missing or unbuildable dependency chain was: ['core-image-minimal-initramfs', 'virtual/kernel']
> ```

ah, sorry, the `poky-tiny` distro actually forces a specific Yocto kernel [1] so it is not meant to be used here — I think I overrode it in the Yocto source for my own testing.

Anyway, switching the target resolved the `busybox-mdev` issue so could you switch back to `DISTRO=poky` and use the same target to build?

[1] https://github.com/yoctoproject/poky/blob/master/meta-poky/conf/distro/poky-tiny.conf#L44


```
meta-poky/conf/distro/poky-tiny.conf#L44
---
# Distro config is evaluated after the machine config, so we have to explicitly
# set the kernel provider to override a machine config.
PREFERRED_PROVIDER_virtual/kernel = "linux-yocto-tiny"
PREFERRED_VERSION_linux-yocto-tiny ?= "6.10%"
```

taraschornyiplv commented:
switching back to poky 
```
Loading cache: 100% |####################################################################################################################################################| Time: 0:00:00
Loaded 3282 entries from dependency cache.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "2.9.1"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-poky-linux"
MACHINE              = "delta-tn48m"
DISTRO               = "poky"
DISTRO_VERSION       = "5.1"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "master:b9b59b5a8bdfa2666477018e7175c5036376807c"
meta-oe              = "master:01dd622533957d20fa860e34ff098161a103bfbf"
meta-switchbsp       = "develop-outoftree-prestera:c0ed18139c9d3a8844602270b008ea203829346b"

Sstate summary: Wanted 1031 Local 1023 Mirrors 0 Missed 8 Current 769 (99% match, 99% complete)###########################################################               | ETA:  0:00:00
Initialising tasks: 100% |###############################################################################################################################################| Time: 0:00:02
NOTE: Executing Tasks
ERROR: core-image-tiny-initramfs-1.0-r0 do_rootfs: Could not invoke dnf. Command '/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/recipe-sysroot-native/usr/bin/dnf -v --rpmverbosity=info -y -c /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/etc/dnf/dnf.conf --setopt=reposdir=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/etc/yum.repos.d --installroot=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs --setopt=logdir=/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp --repofrompath=oe-repo,/home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/oe-rootfs-repo --nogpgcheck install base-passwd busybox busybox-mdev dropbear initramfs-live-boot-tiny packagegroup-core-boot run-postinsts' returned 1:
DNF version: 4.21.1
cachedir: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/rootfs/var/cache/dnf
Added oe-repo repo from /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/oe-rootfs-repo
User-Agent: falling back to 'libdnf': could not detect OS or basearch
repo: using cache for: oe-repo
oe-repo: using metadata from Wed 04 Dec 2024 04:34:16 PM UTC.
No match for argument: busybox-mdev
Error: Unable to find a match: busybox-mdev

ERROR: Logfile of failure stored in: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp/log.do_rootfs.199808
ERROR: Task (/home/taras/git/poky/meta/recipes-core/images/core-image-tiny-initramfs.bb:do_rootfs) failed with exit code '1'
NOTE: Tasks Summary: Attempted 3712 tasks of which 3686 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/taras/git/poky/meta/recipes-core/images/core-image-tiny-initramfs.bb:do_rootfs
    log: /home/taras/git/poky/build/tmp/work/delta_tn48m-poky-linux/core-image-tiny-initramfs/1.0/temp/log.do_rootfs.199808
Summary: There was 1 ERROR message, returning a non-zero exit code.
```

Commit:
conf: common.inc: remove erroneous INITRAMFS_IMAGE forcing

This removes a temporary INITRAMFS_IMAGE variable which would explicity
build a `core-image-tiny-initramfs` variant of the initramfs image
regardless of the actual requested build target.

Reported-by: Taras Chornyi <taras.chornyi@plvision.eu>
Signed-off-by: Jakov Petrina <jakov.petrina@sartura.hr>

jpetrina commented:
Thanks @taraschornyiplv , I went ahead and reproduced this locally. This was indeed an issue in the layer code: a temporary variable was set which would force a specific variant of the initramfs target. 

I've went ahead and pushed be13268bcc446c7a6f3d5443c7f10681be167630 as a fix, let me know if it works for you.
